### ISSUE TITLE
Add exist() function to check for element until timeout - in development

### ISSUE THREAD

**Issue by mbsryu** [kelaberetiv/TagUI#651]
When automating some web page actions using multiple consecutive "click" steps on button-like elements, the flow appears to exit when it cannot find the button or the item to be clicked next.

As the appearance of the next item can be delayed and/or not appear at all, I've tried to compensate by using the following:

if present('changeJobStatus')
{
click changeJobStatus
wait .2
}
if present('submit')
{
click submit
wait .2
}

**Questions:**
1. Are there specific instances where tagui does not wait by default for the result/response to the first click?
2. Is the step "wait" appropriate when it seems that the next click is happening too fast? or does it override the default and shorten the wait time instead?
3. What about the use of: if present('element') ? would this also result in the default waiting time being skipped?

**Comment by kensoh** [comment]
1. TagUI will wait by default until the element appears before acting on it, until default timeout value of 10 seconds before error. The auto-wait happens before an element appears for the step, not after the step. Because you won't know how long after a step and click will the UI update itself to show the next thing you are looking at.

2. The wait step has no impact on the auto-wait before error out, that one is controlled using timeout step. If an element is expected to appear but takes long, you can use the timeout step to set to a long time you think is acceptable, before failing the automation.

3. I normally won't use above structure before the present() function will check if the element exists right now at this instance, and if it has not appear yet, it will be missed.

**Comment by kensoh** [comment]
A situation where your structure above is useful is when you want to handle scenarios differently depending if something appears or not. Then you can do something like if ... do this, else .. do that. And in that situation, present() is not so good. You can try copying below code and save it as tagui_global.js to put in the tagui\src folder. This function is similar to present(), except that it will auto-wait for the timeout duration until the element appears, before returning true or false.

```js
function exist(element_identifier) {

    var exist_timeout = Date.now() + casper.options.waitTimeout;

    while (Date.now() < exist_timeout) {
        if (present(element_identifier))
            return true;
        else
           sleep(100);
    }

    return false;

}
```

Fyi @siowyisheng above function I wrote it for use in TagUI for Python project. But I didn't add it into TagUI because I'm not sure if it will be useful here. Copying for your review to see if this is something interesting that you'll want to add this in v6.

**Comment by siowyisheng** [comment]
Yea it seems useful and common. We should allow an argument for timeout with natural language, then that API would cover both this case and ```present()```. I'll think about the API.

**Comment by kensoh** [comment]
Ok great! For the timeout I'll prefer using timeout step to control the value for the whole automation sequence instead of putting as a parameter. The reason for this preference is if we add timeout as parameter for this, then why not add timeout as a parameter for all the remaining UI automation steps, so can easily go into rabbit hole.

Rationale is bit like if you look at programming languages, the keywords usually try to cover very basic building blocks that users combine to build complex stuffs, but don't add additional layers that do what other keywords do.

But in this case it may be different because TagUI is part domain specific language and part create to make it convenient for users - so you see things like save step which contradicts with the no overlap idea. Because save is actually can be done with read followed by dump step.

So I'll defer the choice to you, since v6 is about moving ahead instead of being held back by 100% backward compatibility!

**Comment by siowyisheng** [comment]
I strongly think it should also have an optional argument.

This is from my experience previously doing a full test suite with the excellent Cypress library, because it was a feature I often used.

**Comment by kensoh** [comment]
Great, let's go for it! I can see the value of having very personalised timeout for different UI steps, and being presented in 1 line instead of having to type 3 lines base on current design.

I've chewed over it, implementation is not that diff even for current architecture, there is a way to implement without changing all the steps and functions. For eg below -

```
click submit with timeout 30 seconds
```

In tagui_parse.php, after creating the .raw file which expands the submodules, code can be added there. To scan for the string ' with timeout '. And convert those lines to -

```
timeout 30 seconds
click submit
timeout 10 seconds
```

This would result in the .raw being parsed with the effect of having custom timeout without having to update all the step definitions and parsing there. Above is still rough, for eg, need to store default timeout somewhere to use and restore back, instead of assuming it is a 10 seconds currently.

And selection criteria needs to be stricter to apply only to relevant TagUI steps, for eg to check the front of the string whether it is an appropriate step for this option. Otherwise if I define a variable string with that string inside, will trigger a match and mess up script.

**Comment by mbsryu** [comment]
I found your response, and the subsequent discussions helpful in understanding how tagui behaves in these situations. I've rid my flow of most of  "wait" steps, and the use of  "if present()" steps to when I have a useful "else" steps to choose from.

**Comment by mbsryu** [comment]
Also the "exist()" step is cool"

**Comment by kensoh** [comment]
@siowyisheng for this issue, shall we scope it just to include the exist() function in v6? If yes I can create a PR so that you can review and see the entry point for this new function.

**Comment by kensoh** [comment]
I will self-assign issues pending my action so it's easier to see what to work on, but probably remove my assignment after i'm done.

**Comment by kensoh** [comment]
As discussed @siowyisheng , scoping the change to adding exist() function for v6, and updating title.

**Comment by kensoh** [comment]
Adding below to tagui_header.js
```js
// present() function that waits until timeout before returning result
function exist(element_identifier) {var exist_timeout = Date.now() + casper.options.waitTimeout;
while (Date.now() < exist_timeout) {if (present(element_identifier)) return true; else sleep(100);}; return false;}
```

**Comment by kensoh** [comment]
raised PR to merge into dev, unassigning myself

**Comment by siowyisheng** [comment]
`exist()` function implemented in #675 for v6
